[IV-24-25] Objetivo 4

ChinChainis opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

ChinChainis committed:
Añadida descripción del proyecto al readme

Añadida información del readme de objetivos 0 y 3 comprendiendo descripción y mantenimiento del proyecto

ChinChainis committed:
Repaso de enlaces del readme

ChinChainis committed:
Cambio de estructura de documentos

ChinChainis committed:
Eliminadas imágenes sobre seguridad del repositorio de raíz de documentos del proyecto

ChinChainis committed:
#24 creación del archivo de estudio del test runner a emplear

ChinChainis committed:
#24 elección de criterios para el test runner a usar

ChinChainis committed:
#24 Añadido estudio de go testing

ChinChainis committed:
#24 Añadido apartado de testify

ChinChainis committed:
#24 Añadido apartado de ginkgo

ChinChainis committed:
#24 Añadido apartado de Maelstrom

ChinChainis committed:
#24 escogemos testing como test runner del proyecto

ChinChainis committed:
#24 cambio de la función esHoraValidaPorTurno a pública

Por temas de sintaxis de go para testearla en otro fichero necesitamos que empiece por mayúscula

ChinChainis committed:
#27 función de testeo de comprobación de turno de las horas

ChinChainis committed:
#26 criterios a estudiar sobre las posibles aserciones

ChinChainis committed:
#26 testify como biblioteca de aserciones

ChinChainis committed:
#26 ghost como biblioteca de aserciones

ChinChainis committed:
#26 assert como biblioteca de aserciones

ChinChainis committed:
#26 conclusión de la biblioteca de aserciones

A menos que haya conflicto con alguna función, emplearemos a poder ser las herramientas nativas de Go

ChinChainis committed:
#27 cambio del struct para múltiples datos de testeo

ChinChainis committed:
#27 ejemplos de uso de la función esHoraValidaPorTurno para su testeo

ChinChainis committed:
#27 reestructurado para comprobar múltiples resultados de la función

ChinChainis committed:
#27 arreglado fallo de tipo de dato en los test

Ejecución correcta de los test, para comprobar escribir línea de go test -v

ChinChainis committed:
#28 función dentro de entidad de horario para asegurar no repeticion

Una misma asignatura no debería estar presente por la mañana y por la tarde a la vez, dado que podría tratarse que el proceso de selección ha escogido la misma asignatura de varios horarios distintos en un mismo resultado

ChinChainis committed:
#28 función para comprobar lógica dentro de creación del objeto horario

ChinChainis committed:
#28 repaso de condicional para función de repaso de lógica de las asignaturas repetidas

ChinChainis committed:
#28 descarte de criterio de testeo

ChinChainis committed:
#24 añadido enlace directo desde el readme

ChinChainis committed:
#24 añadido uso del test runner al justfile

ChinChainis committed:
#24 añadidas instrucciones de testeo al readme

ChinChainis committed:
#24 reestructuración del readme para test

ChinChainis committed:
#24 actualizado iv.yaml para test

ChinChainis committed:
#24 actualizado identacion de iv.yaml

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,25 @@
+
+Es un paquete para proyectos en Go basado en el paquete de assert de testify. Para su uso se necesita importar al proyecto con `import{"github.com/negrel/assert"}` y fue revisada hace un mes desde la escritura de este apartado del documento.
+
+**De momento vamos a emplear go testing.**
```
Hablas de go testing, pero no aparece entre las opciones disponibles no?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/2868d4d97d7ba14b4b5e6e37bdfcee65a20587ce
Se me ha debido colar porque quería la misma estructura del test runner. Además antes de que pueda dar lugar a otra revisión, no sé qué he hecho con push que me ha juntado los dos últimos commits sobre hablar del uso en la comunidad de ghost y el cambio de la conclusión de qué biblioteca de aserciones emplear, soy consciente de que los commits deben ser atómicos no entiendo por qué se han juntado ahí.

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,25 @@
+
+Es un paquete para proyectos en Go basado en el paquete de assert de testify. Para su uso se necesita importar al proyecto con `import{"github.com/negrel/assert"}` y fue revisada hace un mes desde la escritura de este apartado del documento.
+
+**De momento vamos a emplear go testing.**
```
Has elegido go testing, pero como tal no esta explicado, esta mas o menos explicado en la conclusión,pero por ejemplo no se habla nada del cirterio de si esta mantenido o no.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/2868d4d97d7ba14b4b5e6e37bdfcee65a20587ce
Se me ha debido colar porque quería la misma estructura del test runner. Además no sé qué he hecho con push que me ha juntado los dos últimos commits sobre hablar del uso en la comunidad de ghost y el cambio de la conclusión de qué biblioteca de aserciones emplear, soy consciente de que los commits deben ser atómicos no entiendo por qué se han juntado ahí.

mmnuria commented on models/franjahoraria.go:
```diff
@@ -63,7 +63,7 @@ func NewFranjaHoraria(dia DiaSemana, inicio HoraMinuto, duracion Duracion, asign
 }
 
-func esHoraValidaPorTurno(hora int, turno Turno) bool {
+func EsHoraValidaPorTurno(hora int, turno Turno) bool {
```
El nombre de esta funcion crees que es correcto? ese nombre sugiere a que es un test que verifica si una hora es valida, pero, segun veo, no hace eso ¿no?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/2921299b7dca70279a6b1a0df258bded48df433f

mmnuria commented on models/franjahoraria_test.go:
```diff
@@ -0,0 +1,29 @@
+	"testing"
+)
+
+func TestEsHoraValidaPorTurno(t *testing.T) {
```
Los test deben de ser unitarios, ¿no crees que puede ser interesante realizar diferentes test recogiendo cada una de las casuisticas posibles? Realmente estos test, deben comprobar que la implementacion de la logica de negocio del primer milestone es correcta. ¿Hay alguno que verifique incompatibilidad horaria?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/c454a0b6f277ff62d694c09c43e863b94b690b1d

lmchaves commented on models/franjahoraria.go:
```diff
@@ -63,7 +63,7 @@ func NewFranjaHoraria(dia DiaSemana, inicio HoraMinuto, duracion Duracion, asign
+func EsHoraValidaPorTurno(hora int, turno Turno) bool {
 	switch turno {
 	case Manana:
 		return hora >= 8 && hora < 14
```
hay que evitar el uso de números mágicos

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/a91b90ca6830e7c7c4a53319995d75bd885cc43c

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,25 @@
+# Biblioteca de aserciones
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; estudiaremos el estado de los repositorios de las cada herramienta para ver su seguimiento y descartaremos aquellos con más de 6 meses de inactividad.
```
¿Como estudiaras el estado de los repositorios? ¿Que criterio seguirás para establecer que llevan más de 6 meses de inactividad?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/2de84bc07b4dde71a3e44f132b79ab3c2a172fab

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,31 @@
+# Test runners
+
+## Criterios de elección
```
Los criterios son los mismo que los establecidos para la biblioteca de asersiones. Podrias darle una vuelta a eso si te parece.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/72cff576710f941bf1619594e04cb8def38bb748
He decidido cambiar los de la biblioteca de aserciones

lmchaves reviewed (commented):
Buenas se ma ha asignado corregir tu PR, por lo general lo veo bastante completo, he hecho unas puntualizaciones sencillas de resolver.

mmnuria reviewed (commented):
Buenas, se me ha asignado como revisor en tu PR, te he dejado una serie de puntos y aspectos a mejorar. Animo! ya casi está.

Por otro lado, revisa los errores frecuentes ya que, he echado un ojo a los issues y me he encontrado que alguno no sigue las buenas prácticas para superar el objetivo.

ChinChainis committed:
Añadidas variables constantes de horas de entrada y salida

ChinChainis committed:
Sustitución de números mágicos

ChinChainis committed:
#27 renombramiento de función a testear

ChinChainis committed:
#27 revisión de renombramiento de la función

ChinChainis committed:
#26 revisión de criterios a emplear para elección de biblioteca de aserciones

ChinChainis committed:
#26 replanteamiento de pros de testify como biblioteca de aserciones

ChinChainis committed:
#26 reescritura de ghost como biblioteca de aserciones

ChinChainis committed:
#26 reescritura de assert como biblioteca de aserciones

ChinChainis committed:
#26 añadidos más criterios de elección de biblioteca de aserciones

ChinChainis committed:
#26 añadida discusión sobre el uso de assert como herramienta de aserciones

ChinChainis committed:
#26 cambio de conclusión sobre biblioteca de aserciones

ChinChainis committed:
#27 renombramiento del test

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido ninguna revisión o commit nuevo en más de 6 meses.
+- Debe ser compatible con el test runner asignado, en este caso es la herramienta nativa del lenguaje go testing.
+- Evitar aquellas que empleen múltiples instalaciones de librerías externas, daremos prioridad a aquellas optimizadas para el lenguaje go para evitar posible deuda técnica que genere las incompatibilidades.
+- Que cuente con una comunidad activa, el uso extendido de una herramienta permite encontrar rápidamente si hay algún fallo o discutir entre varios usuarios cómo solventarlo.
```
Este cirterio no es medible

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/db6d70c32ab2cc56e2bd0bbe365a736412697814

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/db6d70c32ab2cc56e2bd0bbe365a736412697814

ChinChainis committed:
#26 eliminación de criterio de comunidad

Criterio no medible

ChinChainis committed:
#26 actualizado go.mod para emplear biblioteca de aserciones de testify

ChinChainis committed:
#26 añadido archivo go.sum para seguimiento de paquetes o librerías de las que dependemos

ChinChainis committed:
#26 uso de herramienta assert.equal para testeo de la función de franjahoraria

ChinChainis committed:
#27 añadido test para asegurar límites del horario

ChinChainis committed:
#31 añadida función a grupo.go para comprobar estado de superposición entre horarios de asignaturas según su hora de inicio

ChinChainis committed:
#31 revisión de los parámetros de la función de superposición

ChinChainis committed:
#31 añadida función al constructor

ChinChainis committed:
#31 creado archivo de test de grupo

ChinChainis committed:
#31 cambio de parámetros de entrada de la función de superposición

ChinChainis committed:
#31 creación de función de testeo sobre función de superposición

ChinChainis committed:
#31 especialización de los test de función de superposición

Para el caso de que una asignatura comprenda enteramente a una asignatura en su horario total

ChinChainis committed:
#31 especialización de los test de función de superposición

Para el caso de que una asignatura pise un poco el horario de otra pero no completamente

ChinChainis committed:
#31 especialización de los test de función de superposición

Para el caso de funcione correctamente sin superposición, en distintos turnos incluso

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+# Biblioteca de aserciones
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido ninguna revisión o commit nuevo en más de 6 meses.
```
El critero "mas de 6 meses" es algo arbitrario. Podrías revisar mi PR del objetivo 3 en el cual tuve ese mismo problema.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/ce90c2e869fec5ce41f44dcd244051833d9dd15a
Lo hablé con el profesor y la medición de 6 meses sí es un número arbitrario pero el criterio de actividad de un repositorio de una herramienta es bastante útil y objetivo para ya debatir si queremos una herramienta sobre otra en nuestro proyecto

mmnuria commented on README.md:
```diff
@@ -1,9 +1,28 @@
 # IV_AntonioSanchezAlcaraz_2425
 Repositorio para la asignatura de Infraestructura Virtual
 
+## *Aplicación de administración de horarios*
```
Creo recordar que esto no forma parte de este objetivo, porfavor, trata de corregirlo para que no aparezca como cambio.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/91d6ace7790dfec9ba99a95426e55f30293cad21

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+### Testify
+**Según el [repositorio oficial](https://github.com/stretchr/testify)**
+
+Testify es una serie de paquetes de uso bastante extendido poseyendo uno específico para el testeo con aserciones llamado `assert package`. Este paquete es revisado regularmente según el seguimiento de commits de su repositorio y testify es completamente orientado a lenguaje Go.
```
¿No crees que puede ser más apropiado hablar directamente del paquete `assert` en lugar de toda la biblioteca `testify`?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/f38ad0affd5c28b45f61d84332a35b3675f5ccb3
Estaba hablando del paquete en todo momento, he añadido más información.

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+### Ghost
+**Según el [repositorio oficial](https://github.com/rliebz/ghost)**
+
+Sigue una orientación al lenguaje natural, tiene múltiples palabras reservadas para intentar mejorar de forma intuitiva qué emplea en cada momento, necesita de incluir la librería en el proyecto; tiene un arduo seguimiento, hace 2 semanas fue revisado, así que está en constante actualización en caso de encontrarse algun error.
```
No entiendo bien a lo que te refieres con "necesita de incluir la librería en el proyecto;" ¿Cuando no es necesario incluirla? Creo que es posible que hayas confundido librería con dependencias.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/8e533125b83de2adbfd16104a869405ad307c61d
Sí, he confundido los términos disculpa

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+
+Sigue una orientación al lenguaje natural, tiene múltiples palabras reservadas para intentar mejorar de forma intuitiva qué emplea en cada momento, necesita de incluir la librería en el proyecto; tiene un arduo seguimiento, hace 2 semanas fue revisado, así que está en constante actualización en caso de encontrarse algun error.
+
+### Assert
```
¿Estas seguro de que no es el mismo que el que proporciona `testify`?

ChinChainis replied:
No, es una herramienta inspirada en el paquete assert de Testify pero es un paquete distinto

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+
+Es un paquete para proyectos en Go basado en el paquete de assert de testify pero más especializado. Necesita importar al proyecto con `import{"github.com/negrel/assert"}` y fue revisada hace un mes desde la escritura de este apartado del documento.
+
+**De momento vamos a emplear go testify.**
```
Realmente no estas eligiendo testify, si no su paquete assert ¿o me equivoco?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/c1ef54c9682454fadfc47ace9e5b39b8f6ebaf5f

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,31 @@
+# Test runners
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; estudiaremos el estado de los repositorios de las cada herramienta para ver su seguimiento y descartaremos aquellos con más de 6 meses de inactividad.
```
mismo comentario que antes, "mas de 6 meses" es un criterio arbitrario. Cuidado con eso.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/2cb5dc98650a5db897707c1e222b63136e7d87b3
El tiempo de 6 meses sí es arbitrario pero que un repositorio esté activo o revisado periódicamente es un criterio objetivo y útil para el desarrollo.

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,31 @@
+### Go testing
+**Según la página de documentación oficial de [golang](https://pkg.go.dev/testing) y su [repositorio](https://github.com/golang/go/blob/master/src/testing/testing.go)**
+
+Go testing es una función de un paquete nativo del lenguaje Go para testeo automático de código. Es actualizado regularmente dado que va directamente asociado al lenguaje y no necesita de instalación, para su uso solo hay que importar "testing" en el archivo a usar.
```
La descripcion realmente podemos acceder a ella a traves de la documentación aportado, ¿no crees que aporta más que menciones si cumple con tus criterios?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/02af6927b17d2c22ef66917fd7dd660c4c73e4b9
Hablo en él ya de los criterios establecidos, es cierto que el tema de la instalación es irrelevante.

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,31 @@
+### Testify
+**Según el estudio de [google trends](https://trends.google.com/trends/explore?date=today%205-y&q=golang%20testify,golang%20goconvey,golang%20ginkgo,golang%20httpexpect,golang%20gomega) y su [repositorio](https://github.com/stretchr/testify)**
+
+Testify es una serie de paquetes de uso bastante extendido, tiene una instalación sencilla con solo escribir `$go get github.com/stretchr/testify` y comprende 5 paquetes; respecto a su seguimiento es revisado mensualmente para mantener su uso correcto.
```
Estas hablando de test runner, pero no mencionas nada al respecto.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/34904b78109ff93819b281fcae826476591b4f29

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,31 @@
+### Maelstrom
+**Según una discusión sobre el tema en [reddit](https://www.reddit.com/r/golang/comments/t29c4d/looking_for_a_test_runner_like_pytest/) y su [repositorio](https://github.com/maelstrom-software/maelstrom)**
+
+Herramienta de testeo de forma aislada en contenedores, para instalarse solo necesita escribirse `$wget -q -O - https://github.com/maelstrom-software/maelstrom/releases/latest/download/cargo-maelstrom-x86_64-unknown-linux-gnu.tgz | tar xzf -` que da lugar a múltiples paquetes para emplear este software. Sin embargo su repositorio es actualizado regularmente, hace escasas horas desde la redacción  ...
```
La forma en como se instala no es necesario realmente, si no, el saber cuales de ellas cumplen o no los criterios escogidos.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/29fa88da7516ac63ea4d4a9ca6f662d7a1c44c1b

mmnuria reviewed (commented):
Lo he revisado y te he dejado unos comentarios.

ChinChainis committed:
#24 eliminación de proceso de instalación de Maelstrom

ChinChainis committed:
#24 eliminación de proceso de instalación de Ginkgo

ChinChainis committed:
#24 eliminación de proceso de instalación de Testify

ChinChainis committed:
#24 revisión de comentario sobre testify

ChinChainis committed:
#24 revisión de comentario sobre go testing

ChinChainis committed:
#24 revisión de criterios

Tiempo de 6 meses establecido de forma arbitraria

ChinChainis committed:
Revisión del readme para el objetivo 4

ChinChainis committed:
#26 revisión de testify, añadida información sobre assert

ChinChainis committed:
#26 revisión de ghost, estudio por su página oficial

ChinChainis committed:
#26 reescritura de conclusión de biblioteca_aserciones.md

ChinChainis committed:
#26 repaso de criterios de biblioteca de aserción

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido actividad de revisión reciente o cuidado con respecto a la actualización del lenguaje Go u otras dependencias a usar.
+- Debe ser compatible con el test runner asignado, en este caso es la herramienta nativa del lenguaje go testing.
```
Ahora que estoy revisando, no es un poco obvio que la bilbioteca sea compatible con go?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/4f44b5feb5dd6d7dd60d82189942327982412fe8

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido actividad de revisión reciente o cuidado con respecto a la actualización del lenguaje Go u otras dependencias a usar.
+- Debe ser compatible con el test runner asignado, en este caso es la herramienta nativa del lenguaje go testing.
+- Evitar aquellas que empleen múltiples instalaciones de librerías externas, daremos prioridad a aquellas optimizadas para el lenguaje go para evitar posible deuda técnica que genere las incompatibilidades.
```
Aqui hablas de deuda técnica y en el primer criterio también

ChinChainis replied:
Los dos pueden ocasionar aumento de deuda técnica qué problema hay?

lmchaves replied:
Me refiero a que como ya has hablado de deuda tecnia aquí me centraría en por ejemplo, funcionalidades nativas que reducen al complejidad del entorno

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/4d5934ce1d0c0ca2efffeb4c290a1f37311e7477

lmchaves reviewed (commented):
Lo he revisado nuevamente y hay unas cosillas sencillas de resolver

ChinChainis committed:
#31 repaso de mensajes de test de superposición de horarios

ChinChainis committed:
#26 revisión de formulación del criterio

Dado que empleamos como test runner la función del lenguaje go de test, no necesitamos especificar que sea compatible con go, sino que nos permita de forma cómoda a través del lenguaje mostrar errores o fallos en el código o lógica de este sin complicación

ChinChainis committed:
#26 revisión de formulación del criterio

Especificar buscar centralización de la lógica de desarrollo del proyecto para no complicar el funcionamiento de esta con dependencias adicionales dando lugar a un aumento de deuda técnica y proceso de desarrollo e investigación

ChinChainis commented:
@JJ si pudieras revisarlo cuando puedas

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido actividad de revisión reciente o cuidado con respecto a la actualización del lenguaje Go u otras dependencias a usar.
+- Debe ser compatible con el test runner asignado, permitiendo mostrar mensajes de error según la aserción al test.
```
Por favor, no pongáis criterios obvios. Si no fuera compatible, ¿se consideraría siquiera? Además, ¿qué tiene que ver la compatibilidad con los mensajes de error?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/04d0b112a7cdbe1b9e62b2c4c173be4a6065cd2b

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/daf1d5af8f8d77cd7ec81dcc407fe51826f7bc9f

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,27 @@
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; descartaremos aquellos cuyos repositorios no hayan tenido actividad de revisión reciente o cuidado con respecto a la actualización del lenguaje Go u otras dependencias a usar.
+- Debe ser compatible con el test runner asignado, permitiendo mostrar mensajes de error según la aserción al test.
+- Evitar aquellas que empleen múltiples instalaciones de librerías externas, daremos prioridad a aquellas optimizadas para el lenguaje go para evitar depender de paquetes o dependencias adicionales.
```
¿Te parece que "múltiple" es objetivo? Que haya que instalar *alguna* sí lo es: o hay que instalarla o no. Sin embargo, ¿múltiple es  una? ¿Dos? ¿35?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/commit/04d0b112a7cdbe1b9e62b2c4c173be4a6065cd2b

JJ commented on docs/docs_seguridad_repositorio/settings_git.png:
Esto puedes borrarlo o hacer lo que estimes conveniente una vez superado el objetivo. En todo caso, no tienen por qué estar en este PR.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/a4f95f63d31e85bddae5250e5eee39682064b45e

JJ commented on models/grupo.go:
```diff
@@ -12,10 +14,38 @@ func NewGrupo(nombre string) (*Grupo, error) {
 	}, nil
 }
+
+func NoSuperposicionDeAsignaturasEnGrupo(horarioEntrada []FranjaHoraria) bool {
```
De acuerdo con la ambigüedad de #31 creas una función que se llama igual y que tiene una altísima complejidad, con 2 if anidados dentro de dos for anidados. El problema es que no estás definiendo si vas a testear la superposición para todo el horario, o en qué momento lo vas a hacer, por ejemplo, cada vez que se proponga una asignatura. Así da la impresión de que lo que haces es generar un horario completo, posiblemente incorrecto, para con posterioridad testear si hay algún error. 
Deberías tratar de formular el problema de forma que sea más sencillo de resolver, y por tanto de testear. Una función con complejidad ciclomática alta en general tiene varias responsabilidades y en todo caso es complicado de testear, porque tendrías que intetar testear todas las posibles ramificaciones del if.

JJ commented on models/grupo.go:
```diff
@@ -12,10 +14,38 @@ func NewGrupo(nombre string) (*Grupo, error) {
 	}
 
+	if NoSuperposicionDeAsignaturasEnGrupo(horario.Horario) {
+		return nil, fmt.Errorf("encontradas asignaturas que se superponen")
```
Los mensajes de error siempre deben ser informativos, no genéricos. Los mensajes genéricos no ayudan a quien los lee a buscar una solución a un problema.
Tampoco entiendo muy bien esto. El programa genera un  horario, ¿y si no está bien falla?
Por otro lado, como se lee esto es si NoSuperposición = True es decir, si no hay superposición, ¿entonces hay un error?

JJ reviewed (changes requested):
Veo una única función aquí que pueda corresponder con la lógica de negocio y por tanto con la HU #2. ¿Te parece que detectar si hay superposiciones y dar un error si es así soluciona el problema de Juan Carlos? ¿Es es suficiente para "ver qué opciones tendría para poner tener un horario con clases en el grupo D y C que son de turno de tarde"? ¿En qué issue está planteado ese problema?

ChinChainis committed:
#26 redefinición de criterio

La palabra múltiple puede ser subjetiva, debe ser un estado binario o se necesitan instalar librerías o no

ChinChainis committed:
#26 eliminado criterio de compatibilidad

Resulta obvio porque si no lo fuera no sería compatible con el lenguaje Go, siendo irrelevante

ChinChainis committed:
Eliminados archivos de configuración para este PR

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+
+## Opciones disponibles para aserciones:
+
+### Testify
```
Si se están proponiendo opciones que ofrezcan aserciones, ¿porqué se menciona Testify? ¿No crees que es suficiente hablando únicamente del paquete `assert`?

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/89fa5c9a7305ec223c6ff872ab2b660d73b8f9f2

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+**Según el [repositorio oficial](https://github.com/negrel/assert)**
+
+Es un paquete para proyectos en Go basado en el paquete de assert de testify pero más especializado. Necesita importar al proyecto con `import{"github.com/negrel/assert"}` y fue revisada hace un mes desde la escritura de este apartado del documento.
+
```
Para que se vea de manera más fácil, podrías añadir un apartado donde se hable de la "elección final" o algo del estilo, ya que es confuso (parece que esas lineas pertenecen a las explicaciones de `Aseert`).

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/c233db7c96f9f36dde2798363ed51d3fc261fd07

mmnuria replied:
Este comentario sigue sin resolver, porfavor es necesario resolverlo modificando el codigo.

mmnuria commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+### Assert
+**Según el [repositorio oficial](https://github.com/negrel/assert)**
+
+Es un paquete para proyectos en Go basado en el paquete de assert de testify pero más especializado. Necesita importar al proyecto con `import{"github.com/negrel/assert"}` y fue revisada hace un mes desde la escritura de este apartado del documento.
```
Mismo comentario que la anterior revisión. No es necesario explicar como se importan al proyecto.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/c4a10dc1965102fd7f5dc786953145b2da376f6a

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,32 @@
+# Test runners
+
+## Criterios de elección
+- Debe tener un mantenimiento en la actualidad para no aumentar la deuda técnica; estudiaremos el estado de los repositorios de las cada herramienta para ver su seguimiento.
```
¿como vas a estudiar el "estado de los repositorios"? Dale una vuelta porque no es un critero objetivo.

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/58f326060ed18f620bf2b6534944438a78be8ffb

JJ replied:
¿Por qué seguís poniendo esto? Si está outdated porque habéis cambiado la línea donde estaba, está solucionado. Y si no estuviera solucionado, poner el commit en el que se ha "solucionado" no lo va a hacer. De verdad, tratad de entender por qué hacéis las cosas y no generéis 200 mensajes de correo electrónico con notificaciones que no aportan ningún valor a nadie.

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,32 @@
+### Testify
+**Según el estudio de [google trends](https://trends.google.com/trends/explore?date=today%205-y&q=golang%20testify,golang%20goconvey,golang%20ginkgo,golang%20httpexpect,golang%20gomega) y su [repositorio](https://github.com/stretchr/testify)**
+
+Testify es una serie de paquetes de uso bastante extendido, comprende 5 paquetes, cada uno con funcionalidades distintas como aserciones o creación de mockups que a lo mejor no se acaban usando en este proyecto; respecto a su seguimiento es revisado mensualmente para mantener su uso correcto.
```
Se están estudiando test runners y no se menciona nada de test runner en esta opción. Tampoco de la actualizacion de ese "test runner".

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/74c9ecfa42dc2db2280e962d37d1459717429f9d

mmnuria commented on docs/test_runners.md:
```diff
@@ -0,0 +1,32 @@
+
+Herramienta de testeo de forma aislada en contenedores para su uso requiere múltiples paquetes para emplear este software. Sin embargo su repositorio es actualizado regularmente, hace escasas horas desde la redacción de este párrafo que se actualizó.
+
+
```
Igual que antes, sería aconsejable crear un apartado para la decision final (evita confusiones mientras se lee)

ChinChainis replied:
https://github.com/ChinChainis/Proyecto_Reparahorarios_IV2425/pull/30/commits/71b2154f562822c540b7c188de21ecf06c213d2f

mmnuria replied:
Este comentario sigue sin resolver, porfavor es necesario resolverlo modificando el codigo.

mmnuria reviewed (commented):
Porfavor, presta atención a la redacción de los diferentes archivos, hay algunos que no se entienden muy bien y los tengo que leer varias veces para tratar de entenderlos. Sería conveniente que los revises para añadir comas, puntos o conectores en su caso, para que se entienda mejor.

Por otro lado, presta atención a los mensajes de los test por ejemplo: "Los resultados deberían de ser iguales" ¿crees que esta aportando algo? Los mensajes de test deben de ser informativos para poder saber en todo momento el problema y así buscarle una solución.

mmnuria commented:
Por último acabo de revisar los issues y deberías de revisar si de verdad crees que están expresando un problema o simplemente estás refiriéndote a la solución directamente. Revisa el archivo de errores frecuentes, puede que te ayude a verlo más claro.

ChinChainis committed:
#32 creada función de comparación de horarios entre dos asignaturas

ChinChainis committed:
#33 compara dos a dos los horarios de un conjunto de asignaturas

ChinChainis committed:
#33 función para estudiar la compatibilidad de un día concreto

ChinChainis committed:
#33 renombramiento de funciones

ChinChainis committed:
#31 función recorre los días de una semana

Comprueba los días de la semana y da un mensaje de error del día que presente algún horario problemático

ChinChainis committed:
#31 función de testeo de superposición de horarios de asignaturas en una semana

ChinChainis committed:
#26 especificación del paquete assert de testify

ChinChainis committed:
#26 añadido apartado de confusión

Debido al formato markdown parecía ser un párrafo de assert, hay que repasar el resultado del markdown al subirse para asegurarse la distribución de texto elegida

ChinChainis committed:
#26 revisado apartado de assert

ChinChainis committed:
#24 revisión de criterio de actualidad

Eliminar elementos subjetivos, simplemente mirar si el repositorio tiene un mantenimiento regular

ChinChainis committed:
#24 revisión de apartado de testify

Añadida información sobre sus funciones de test runner por medio de introducción de mockups

ChinChainis committed:
#24 añadido apartado de conclusión final a testrunners

Revisar estado final del archivo markdown para evitar confusión

ChinChainis committed:
#33 repaso de mensaje de error de función de comprobar superposiciones de una semana

ChinChainis committed:
#32 renombramiento de función de comparación de dos días

mmnuria reviewed (commented):
He añadido algunos comentarios extra pero me gustaría mencionar varias cosas:

1. Revisa con atención los **issues** creados que se están usando en los commits, ya que, **no expresan ningún problema,** si no, una solución. Es necesario que esto se revise para poder superar el objetivo.
2. Presta atención ya que hay 3 issues (31,32 y 33) que usas para el objetivo y **no tienen asociado ningún milestone** (esto es un requerimiento de la asignatura).
3. Los comentarios se deben resolver haciendo **cambios en el código**, porfavor revisa los que estén abiertos y continuaré con mi revisión. No es necesario incluir en el comentario ningún commit asociado, **con hacer cambios en la misma linea en la que se hace el comentario** es suficiente.

ChinChainis committed:
#24 revisión de conclusión final

ChinChainis committed:
#24 revisión de apartado de Maelstrom

ChinChainis committed:
#24 revisión de apartado de Maelstrom

Añadido enlace a página web oficial

ChinChainis committed:
#24 revisión estructura del archivo

ChinChainis committed:
#24 revisión estructura del archivo espacios entre párrafos

ChinChainis committed:
#24 revisión estructura del archivo espacios entre párrafos

ChinChainis committed:
#24 revisión estructura del archivo espacios entre párrafos

ChinChainis committed:
#24 revisión estructura saltos de página

ChinChainis committed:
#26 añadida información de assert

ChinChainis committed:
#26 cambio formato archivo biblioteca de aserciones

ChinChainis committed:
#26 cambio paquete assert y conclusión final

ChinChainis committed:
#24 actualizado apartado de maelstrom

ChinChainis committed:
#24 actualizado conclusión

ChinChainis committed:
#24 actualizado formato maelstrom y conclusión

ChinChainis committed:
#24 actualizado formato archivo test_runners

ChinChainis committed:
#24 actualizado espaciado test runners

ChinChainis committed:
#24 actualizado formato test runners

ChinChainis committed:
#24 actualizado formato test runners

ChinChainis committed:
#34 función de creación de un horario dado un día

ChinChainis committed:
#34 función de testeo de un horario por día funcional

ChinChainis committed:
#34 renombramiento de la función

ChinChainis committed:
#34 actualización del nombre ne función de testeo

ChinChainis committed:
#34 actualizado comentario de función

ChinChainis committed:
#34 corrección de función assert del test de haceHorarioDeUnDia

ChinChainis committed:
#34 especificación del mensaje de error

ChinChainis committed:
#35 añadida función a grupo de generación de horario en una semana

ChinChainis committed:
#31 renombramiento de variables de función de superposición en una semana

ChinChainis committed:
#35 función de slice para generar horario resultante

ChinChainis committed:
#34 revisión de función de un horario de un día

Mejor división del trabajo para evitar complejidad ciclomática

ChinChainis committed:
#35 añadido test para generar horario de una semana

JJ commented:
@mmnuria ¿puedes mirar los cambios hechos? He visto que todavía no has aprobado. Si no puedes dímelo y paso yo a revisarlo.

mmnuria commented on docs/test_runners.md:
```diff
@@ -27,5 +27,4 @@ Conjunto de herramientas de ejecución de test de forma aislada en micro-contene
 Go testing al ser una función parte del propio lenguaje es actualizada junto al lenguaje asegurando que no quede desfasada, es la herramienta más directa y no requiere dependencias externas; permitiendo instalación mínima para el testeo automático de los archivos del proyecto. 
-
-**Vamos a emplear go testing como test runner.**
+Por ello: **Vamos a emplear go testing como test runner.**
```
¿Es go testing un test runner? Creo que eso lo habíamos hablado antes.

ChinChainis replied:
He declarado que es un paquete que sirve para testear código y que lo vamos a emplear como test runner, no he dicho que sea un test runner es solo un framework de testeo

mmnuria reviewed (commented):
Te dejo unos comentarios extra:

1. Revisa en algunas funciones test para evitar la complejidad ciclomátia (if dentro de for)
2. Los nombres de las funciones no son del todo descriptivos y no logro entender bien para que se usan, intenta ver como se puede mejorar esa parte.
3. Revisa los issues, hay algunos que no expresan un problema, si no una solución.

mmnuria commented:
> @mmnuria ¿puedes mirar los cambios hechos? He visto que todavía no has aprobado. Si no puedes dímelo y paso yo a revisarlo.

Revisado.

ChinChainis committed:
Actualización del gitignore

ChinChainis committed:
Actualización del gitignore

ChinChainis committed:
#24 reformulación de conclusión final

ChinChainis committed:
#33 repaso de función de superposición en una semana

ChinChainis committed:
#34 unificación de funciones para simplificar lógica de creación de horario en un día

ChinChainis committed:
#33 revisión de función de comprobar superosiciones en semana

ChinChainis committed:
#33 cambio de función a emplear para comprobar superposicion

ChinChainis committed:
#34 simplificación de función de crear un horario de un día

ChinChainis committed:
#32 función replanteada para evitar complejidad ciclomática

ChinChainis committed:
#33 arreglada lógica de las funciones que comprueba superposicion en semana

ChinChainis committed:
#34 revisión de funciones para generar de forma lógica un horario

ChinChainis committed:
#34 revisión de test

ChinChainis committed:
#31 renombramiento de las funciones de revisión de superposiciones

ChinChainis commented:
@mmnuria Entiendo la queja de la complicidad ciclomática y la he reducido lo máximo posible pero hay solo un par de funciones que debido a las características de golang tengo que recorrerlo con bucles para cumplir su cometido, no es un tema de slicing es ya de búsqueda de valores en un array y no hay otra forma en golang.

mmnuria commented:
@ChinChainis Por mi parte tienes el apto. Muy buen trabajo!

mmnuria reviewed (approved):
@ChinChainis Por mi parte tienes el apto. Muy buen trabajo!

ChinChainis commented:
@JJ si quieres echarle un vistazo

JJ commented on models/franjahoraria_test.go:
```diff
@@ -0,0 +1,59 @@
+	"github.com/stretchr/testify/assert"
+)
+
+func TestCorrespondenciaHoraTurno(t *testing.T) {
```
Los tests deben resolver el problema del cliente. Si simplemente comprueban si las funciones hacen alguna operación correctamente, sólo añaden ruido y problemas a la revisión. Estás comprobando esencialmente si la asignación en Go funciona bien, por otro lado.

JJ reviewed (commented):
Te hago un pequeño comentario por si pudiera aportarte algo, aunque debido a que no has entregado el objetivo 5 a tiempo, tampoco es que sea necesario. Si quieres seguir trabajando en este objetivo, te lo sigo revisando sin problemas, incluso más allá de la fecha de entrega. Siento que no haya sido posible que superes los objetivos mínimos de la asignatura.
